{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyOsZXntqLptIlDDkL78M5o8",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/Jayalakshmi-27-05/Bioinformatics_lab/blob/main/7_genome_assembly.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "vwrEgPIZD8Iw",
        "outputId": "126c4d91-6423-4217-c800-4ef6544d9468"
      },
      "source": [
        "!apt-get install fastqc\n",
        "!apt-get install velvet\n",
        "!wget 'https://zenodo.org/record/582600/files/mutant_R1.fastq'\n",
        "!wget 'https://zenodo.org/record/582600/files/mutant_R2.fastq'\n",
        "import os\n",
        "from google.colab import files\n",
        "get_ipython().system(\"fastqc mutant_R1.fastq\")\n",
        "files.download('mutant_R1_fastqc.html')\n",
        "get_ipython().system(\"fastqc mutant_R2.fastq\")\n",
        "files.download('mutant_R2_fastqc.html')\n",
        "!mkdir assembly_module\n",
        "os.chdir('assembly_module')\n",
        "!mkdir data\n",
        "get_ipython().system(\"velveth ASM_NAME 31 -short -separate -fastq /content/mutant_R1.fastq /content/mutant_R2.fastq\")\n",
        "get_ipython().system(\"velvetg ASM_NAME\")"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "fastqc is already the newest version (0.11.9+dfsg-5).\n",
            "0 upgraded, 0 newly installed, 0 to remove and 67 not upgraded.\n",
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "velvet is already the newest version (1.2.10+dfsg1-7).\n",
            "0 upgraded, 0 newly installed, 0 to remove and 67 not upgraded.\n",
            "--2025-12-09 04:39:14--  https://zenodo.org/record/582600/files/mutant_R1.fastq\n",
            "Resolving zenodo.org (zenodo.org)... 188.185.43.153, 188.185.48.75, 137.138.52.235, ...\n",
            "Connecting to zenodo.org (zenodo.org)|188.185.43.153|:443... connected.\n",
            "HTTP request sent, awaiting response... 301 MOVED PERMANENTLY\n",
            "Location: /records/582600/files/mutant_R1.fastq [following]\n",
            "--2025-12-09 04:39:14--  https://zenodo.org/records/582600/files/mutant_R1.fastq\n",
            "Reusing existing connection to zenodo.org:443.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 4137809 (3.9M) [application/octet-stream]\n",
            "Saving to: ‘mutant_R1.fastq.3’\n",
            "\n",
            "mutant_R1.fastq.3   100%[===================>]   3.95M   902KB/s    in 4.5s    \n",
            "\n",
            "2025-12-09 04:39:19 (902 KB/s) - ‘mutant_R1.fastq.3’ saved [4137809/4137809]\n",
            "\n",
            "--2025-12-09 04:39:19--  https://zenodo.org/record/582600/files/mutant_R2.fastq\n",
            "Resolving zenodo.org (zenodo.org)... 188.185.43.153, 188.185.48.75, 137.138.52.235, ...\n",
            "Connecting to zenodo.org (zenodo.org)|188.185.43.153|:443... connected.\n",
            "HTTP request sent, awaiting response... 301 MOVED PERMANENTLY\n",
            "Location: /records/582600/files/mutant_R2.fastq [following]\n",
            "--2025-12-09 04:39:20--  https://zenodo.org/records/582600/files/mutant_R2.fastq\n",
            "Reusing existing connection to zenodo.org:443.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 4137809 (3.9M) [application/octet-stream]\n",
            "Saving to: ‘mutant_R2.fastq.3’\n",
            "\n",
            "mutant_R2.fastq.3   100%[===================>]   3.95M   819KB/s    in 4.9s    \n",
            "\n",
            "2025-12-09 04:39:25 (819 KB/s) - ‘mutant_R2.fastq.3’ saved [4137809/4137809]\n",
            "\n",
            "Started analysis of mutant_R1.fastq\n",
            "Approx 5% complete for mutant_R1.fastq\n",
            "Approx 15% complete for mutant_R1.fastq\n",
            "Approx 20% complete for mutant_R1.fastq\n",
            "Approx 30% complete for mutant_R1.fastq\n",
            "Approx 40% complete for mutant_R1.fastq\n",
            "Approx 45% complete for mutant_R1.fastq\n",
            "Approx 55% complete for mutant_R1.fastq\n",
            "Approx 60% complete for mutant_R1.fastq\n",
            "Approx 70% complete for mutant_R1.fastq\n",
            "Approx 80% complete for mutant_R1.fastq\n",
            "Approx 85% complete for mutant_R1.fastq\n",
            "Approx 95% complete for mutant_R1.fastq\n",
            "Analysis complete for mutant_R1.fastq\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "\n",
              "    async function download(id, filename, size) {\n",
              "      if (!google.colab.kernel.accessAllowed) {\n",
              "        return;\n",
              "      }\n",
              "      const div = document.createElement('div');\n",
              "      const label = document.createElement('label');\n",
              "      label.textContent = `Downloading \"${filename}\": `;\n",
              "      div.appendChild(label);\n",
              "      const progress = document.createElement('progress');\n",
              "      progress.max = size;\n",
              "      div.appendChild(progress);\n",
              "      document.body.appendChild(div);\n",
              "\n",
              "      const buffers = [];\n",
              "      let downloaded = 0;\n",
              "\n",
              "      const channel = await google.colab.kernel.comms.open(id);\n",
              "      // Send a message to notify the kernel that we're ready.\n",
              "      channel.send({})\n",
              "\n",
              "      for await (const message of channel.messages) {\n",
              "        // Send a message to notify the kernel that we're ready.\n",
              "        channel.send({})\n",
              "        if (message.buffers) {\n",
              "          for (const buffer of message.buffers) {\n",
              "            buffers.push(buffer);\n",
              "            downloaded += buffer.byteLength;\n",
              "            progress.value = downloaded;\n",
              "          }\n",
              "        }\n",
              "      }\n",
              "      const blob = new Blob(buffers, {type: 'application/binary'});\n",
              "      const a = document.createElement('a');\n",
              "      a.href = window.URL.createObjectURL(blob);\n",
              "      a.download = filename;\n",
              "      div.appendChild(a);\n",
              "      a.click();\n",
              "      div.remove();\n",
              "    }\n",
              "  "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "download(\"download_35458d04-cb0c-48f3-8ad1-dda79f418e9d\", \"mutant_R1_fastqc.html\", 543669)"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Started analysis of mutant_R2.fastq\n",
            "Approx 5% complete for mutant_R2.fastq\n",
            "Approx 15% complete for mutant_R2.fastq\n",
            "Approx 20% complete for mutant_R2.fastq\n",
            "Approx 30% complete for mutant_R2.fastq\n",
            "Approx 40% complete for mutant_R2.fastq\n",
            "Approx 45% complete for mutant_R2.fastq\n",
            "Approx 55% complete for mutant_R2.fastq\n",
            "Approx 60% complete for mutant_R2.fastq\n",
            "Approx 70% complete for mutant_R2.fastq\n",
            "Approx 80% complete for mutant_R2.fastq\n",
            "Approx 85% complete for mutant_R2.fastq\n",
            "Approx 95% complete for mutant_R2.fastq\n",
            "Analysis complete for mutant_R2.fastq\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "\n",
              "    async function download(id, filename, size) {\n",
              "      if (!google.colab.kernel.accessAllowed) {\n",
              "        return;\n",
              "      }\n",
              "      const div = document.createElement('div');\n",
              "      const label = document.createElement('label');\n",
              "      label.textContent = `Downloading \"${filename}\": `;\n",
              "      div.appendChild(label);\n",
              "      const progress = document.createElement('progress');\n",
              "      progress.max = size;\n",
              "      div.appendChild(progress);\n",
              "      document.body.appendChild(div);\n",
              "\n",
              "      const buffers = [];\n",
              "      let downloaded = 0;\n",
              "\n",
              "      const channel = await google.colab.kernel.comms.open(id);\n",
              "      // Send a message to notify the kernel that we're ready.\n",
              "      channel.send({})\n",
              "\n",
              "      for await (const message of channel.messages) {\n",
              "        // Send a message to notify the kernel that we're ready.\n",
              "        channel.send({})\n",
              "        if (message.buffers) {\n",
              "          for (const buffer of message.buffers) {\n",
              "            buffers.push(buffer);\n",
              "            downloaded += buffer.byteLength;\n",
              "            progress.value = downloaded;\n",
              "          }\n",
              "        }\n",
              "      }\n",
              "      const blob = new Blob(buffers, {type: 'application/binary'});\n",
              "      const a = document.createElement('a');\n",
              "      a.href = window.URL.createObjectURL(blob);\n",
              "      a.download = filename;\n",
              "      div.appendChild(a);\n",
              "      a.click();\n",
              "      div.remove();\n",
              "    }\n",
              "  "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "download(\"download_77bb231e-61fd-4fed-9dd8-2fbffc902350\", \"mutant_R2_fastqc.html\", 547109)"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "mkdir: cannot create directory ‘assembly_module’: File exists\n",
            "[0.000000] Reading FastQ file /content/mutant_R1.fastq;\n",
            "[0.053976] 12480 sequences found\n",
            "[0.054000] Done\n",
            "[0.054053] Reading FastQ file /content/mutant_R2.fastq;\n",
            "[0.115376] 12480 sequences found\n",
            "[0.115395] Done\n",
            "[0.115464] Reading read set file ASM_NAME/Sequences;\n",
            "[0.122075] 24960 sequences found\n",
            "[0.153862] Done\n",
            "[0.153887] 24960 sequences in total.\n",
            "[0.154009] Writing into roadmap file ASM_NAME/Roadmaps...\n",
            "[0.205618] Inputting sequences...\n",
            "[0.205650] Inputting sequence 0 / 24960\n",
            "[1.104092]  === Sequences loaded in 0.898477 s\n",
            "[1.104135] Done inputting sequences\n",
            "[1.104138] Destroying splay table\n",
            "[1.115521] Splay table destroyed\n",
            "[0.000000] Reading roadmap file ASM_NAME/Roadmaps\n",
            "[0.049871] 24960 roadmaps read\n",
            "[0.050148] Creating insertion markers\n",
            "[0.053743] Ordering insertion markers\n",
            "[0.073245] Counting preNodes\n",
            "[0.076988] 64223 preNodes counted, creating them now\n",
            "[0.177720] Adjusting marker info...\n",
            "[0.181471] Connecting preNodes\n",
            "[0.220379] Cleaning up memory\n",
            "[0.220517] Done creating preGraph\n",
            "[0.220526] Concatenation...\n",
            "[0.241792] Renumbering preNodes\n",
            "[0.241825] Initial preNode count 64223\n",
            "[0.244172] Destroyed 37924 preNodes\n",
            "[0.244206] Concatenation over!\n",
            "[0.244210] Clipping short tips off preGraph\n",
            "[0.245889] Concatenation...\n",
            "[0.250954] Renumbering preNodes\n",
            "[0.250984] Initial preNode count 26299\n",
            "[0.252678] Destroyed 3832 preNodes\n",
            "[0.252696] Concatenation over!\n",
            "[0.252698] 1986 tips cut off\n",
            "[0.252700] 22467 nodes left\n",
            "[0.252797] Writing into pregraph file ASM_NAME/PreGraph...\n",
            "[0.326694] Reading read set file ASM_NAME/Sequences;\n",
            "[0.331539] 24960 sequences found\n",
            "[0.354823] Done\n",
            "[0.407328] Reading pre-graph file ASM_NAME/PreGraph\n",
            "[0.407469] Graph has 22467 nodes and 24960 sequences\n",
            "[0.435362] Scanning pre-graph file ASM_NAME/PreGraph for k-mers\n",
            "[0.440899] 434645 kmers found\n",
            "[0.473903] Sorting kmer occurrence table ... \n",
            "[0.608406] Sorting done.\n",
            "[0.608434] Computing acceleration table... \n",
            "[0.666148] Computing offsets... \n",
            "[0.669506] Ghost Threading through reads 0 / 24960\n",
            "[0.669845]  === Ghost-Threaded in 0.000340 s\n",
            "[0.669859] Threading through reads 0 / 24960\n",
            "[2.138576]  === Threaded in 1.468717 s\n",
            "[2.144536] Correcting graph with cutoff 0.200000\n",
            "[2.147322] Determining eligible starting points\n",
            "[2.176208] Done listing starting nodes\n",
            "[2.176242] Initializing todo lists\n",
            "[2.183238] Done with initialization\n",
            "[2.183279] Activating arc lookup table\n",
            "[2.187033] Done activating arc lookup table\n",
            "[2.311665] 10000 / 22467 nodes visited\n",
            "[2.485422] 20000 / 22467 nodes visited\n",
            "[2.528258] Concatenation...\n",
            "[2.528543] Renumbering nodes\n",
            "[2.528558] Initial node count 22467\n",
            "[2.528734] Removed 22097 null nodes\n",
            "[2.528746] Concatenation over!\n",
            "[2.528767] Clipping short tips off graph, drastic\n",
            "[2.528812] Concatenation...\n",
            "[2.529086] Renumbering nodes\n",
            "[2.529098] Initial node count 370\n",
            "[2.529112] Removed 14 null nodes\n",
            "[2.529120] Concatenation over!\n",
            "[2.529127] 356 nodes left\n",
            "[2.529303] Writing into graph file ASM_NAME/Graph...\n",
            "[2.563318] WARNING: NO COVERAGE CUTOFF PROVIDED\n",
            "[2.563349] Velvet will probably leave behind many detectable errors\n",
            "[2.563351] See manual for instructions on how to set the coverage cutoff parameter\n",
            "[2.563355] Removing contigs with coverage < -1.000000...\n",
            "[2.563391] Concatenation...\n",
            "[2.563417] Renumbering nodes\n",
            "[2.563423] Initial node count 356\n",
            "[2.563430] Removed 0 null nodes\n",
            "[2.563436] Concatenation over!\n",
            "[2.563452] Concatenation...\n",
            "[2.563470] Renumbering nodes\n",
            "[2.563476] Initial node count 356\n",
            "[2.563482] Removed 0 null nodes\n",
            "[2.563488] Concatenation over!\n",
            "[2.563494] Clipping short tips off graph, drastic\n",
            "[2.563507] Concatenation...\n",
            "[2.563525] Renumbering nodes\n",
            "[2.563531] Initial node count 356\n",
            "[2.563534] Removed 0 null nodes\n",
            "[2.563540] Concatenation over!\n",
            "[2.563546] 356 nodes left\n",
            "[2.563552] WARNING: NO EXPECTED COVERAGE PROVIDED\n",
            "[2.563557] Velvet will be unable to resolve any repeats\n",
            "[2.563563] See manual for instructions on how to set the expected coverage parameter\n",
            "[2.563569] Concatenation...\n",
            "[2.563587] Renumbering nodes\n",
            "[2.563593] Initial node count 356\n",
            "[2.563599] Removed 0 null nodes\n",
            "[2.563605] Concatenation over!\n",
            "[2.563610] Removing reference contigs with coverage < -1.000000...\n",
            "[2.563634] Concatenation...\n",
            "[2.563653] Renumbering nodes\n",
            "[2.563658] Initial node count 356\n",
            "[2.563664] Removed 0 null nodes\n",
            "[2.563669] Concatenation over!\n",
            "[2.563939] Writing contigs into ASM_NAME/contigs.fa...\n",
            "[2.590429] Writing into stats file ASM_NAME/stats.txt...\n",
            "[2.592279] Writing into graph file ASM_NAME/LastGraph...\n",
            "Final graph has 356 nodes and n50 of 5029, max 12902, total 184065, using 0/24960 reads\n"
          ]
        }
      ]
    }
  ]
}